overlays/nodejs/lambda/lambdaMixins.ts

```diff
@@ -0,0 +1,308 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as pulumi from "@pulumi/pulumi";
+import * as crypto from "crypto";
+
+import * as arn from "../arn";
+import * as iam from "../iam";
+import * as runtime from "./runtimes";
+import * as utils from "../utils";
+
+import * as lambdaFunction from "./function";
+import * as permission from "./permission";
+
+/**
+ * Context is the shape of the context object passed to a Function callback.
+ */
+export interface Context {
+    callbackWaitsForEmptyEventLoop: boolean;
+    readonly functionName: string;
+    readonly functionVersion: string;
+    readonly invokedFunctionArn: string;
+    readonly memoryLimitInMB: string;
+    readonly awsRequestId: string;
+    readonly logGroupName: string;
+    readonly logStreamName: string;
+    readonly identity: any;
+    readonly clientContext: any;
+    getRemainingTimeInMillis(): string;
+}
+
+/**
+ * EntryPoint is the signature for a serverless function that will be invoked each time the AWS
+ * Lambda is invoked.
+ *
+ * It can be a synchronous or asynchronous function that can be converted into an AWS lambda.  Async
+ * callbacks are only supported with an AWS lambda runtime of 8.10 or higher.  On those runtimes a
+ * Promise can be returned, 'callback' parameter can be ignored, and AWS will appropriately handle
+ * things. For AWS lambda pre-8.10, a synchronous function must be provided.  The synchronous
+ * function should return nothing, and should instead invoke 'callback' when complete.
+ */
+export type EntryPoint<E, R> = (event: E, context: Context, callback: (error: any, result: R) => void) => Promise<R> | void;
+
+/**
+ * EntryPointFactory is the signature for a function that will be called once to produce the
+ * serverless function that AWS Lambda will invoke.  It can be used to initialize expensive state
+ * once that can then be used across all invocations of the Lambda (as long as the Lambda is using
+ * the same warm node instance).
+ */
+export type EntryPointFactory<E, R> = () => EntryPoint<E, R>;
```

Comment:
these types were renamed as generic names like Func/Handler were getting quite confusing.  "EntryPoint" clearly indicates (along with teh docs) that these represent the initial entrypoints for a Lambda.  

These have also been made strongly typed, as we had a weak typed and strong typed version of these before.  Now, we'll just have the strong typed version.  If someone wants the weak typed version they can use `Entrypoint<any, any>`

Reply:
How about something like `FunctionCallback` or, to make it look better alongside `EventHandler`, perhaps `EventCallback`.

Reply:
I really like lambda.EntryPoint. It makes it really clear waht it's describing. i.e. it's not the shape of an arbitrary feature-level callback. It's expressing hte shape of hte Lambda entrypoint. It's very lambda specific (since it gets that wacky callback object and has that specific shape). EntryPointFactory also reads well to me.

Reply:
I'll admit I find `EntryPoint` a bit surprising here as well - in part because I've never seen that term used in Lambda docs.  In their docs, this is a "handler" or "callback".  

Note even above where the doc comments say "Context is the shape of the context object passed to a Function callback." and include a property named `callbackWaitsForEmptyEventLoop`.  Given that, I think `FunctionCallback` would feel more natural here.